src/tools.cc

```diff
@@ -971,17 +970,8 @@ parseEtcHosts(void)
             *nt = '\0';
             debugs(1, 5, "etc_hosts: got hostname '" << lt << "'");
 
-            /* For IPV6 addresses also check for a colon */
-            if (Config.appendDomain && !strchr(lt, '.') && !strchr(lt, ':')) {
-                /* I know it's ugly, but it's only at reconfig */
-                strncpy(buf2, lt, sizeof(buf2)-1);
-                strncat(buf2, Config.appendDomain, sizeof(buf2) - strlen(lt) - 1);
-                buf2[sizeof(buf2)-1] = '\0';
-                host = buf2;
-            } else {
-                host = lt;
-            }
-
+            (void)urlAppendDomain(lt);
```

Comment:
Why are we ignoring the false result of the urlAppendDomain() call? Does not it indicate a problem worth handling here?

Reply:
urlAppendDomain is a complete replacement for ifelse. false response is unnecessary at this point.

Reply:
AFAICT, your "we are preserving the old behavior" answer assumes that the old code correctly ignored failures. Let's forget what the (buggy) old code did for a second. Can you explain why ignoring failures is correct in this context? As a Squid administrator, I would expect Squid to _at least_ warn me that my `append_domain` directive was not honored. What's wrong with those expectations?

Reply:
my fix aims to eliminate the overflow as soon as possible. and we have been discussing for 3 days how best to fix it. It seems to me that the correct practice is to quickly fix the vulnerable section of the code and then calmly develop and improve the code with discussion.

Reply:
AFAICT, there is no rush with this fix -- there is no emergency here. Please take your time to "calmly develop" a quality fix. As for the best practices, they vary from project to project, and even from reviewer to reviewer. If you want to know my personal views, you can find them [elsewhere](https://github.com/measurement-factory/squid-notes/blob/start/review.md).

Reply:
FWIW, the append function already handles reporting that domain too long and by how much. The false result is for the caller if it wants to do any further handling eg error, fatal, context message explaining what the domain is being used for, etc.

Reply:
> the append function already handles reporting that domain too long and by how much

FYI: urlAppendDomain() reports nothing to Squid administrators. Its level-2 debugging is irrelevant in the context of this change request.

It is possible that urlAppendDomain() itself should do more than it does now (e.g., throw an exception), but I have not checked all the details to recommend a specific solution. In many cases like this one, it is best to fix the current caller first and then look at whether all callers or the function should be adjusted to unify handling. Both steps should be done in one PR, but splitting this work into two steps may help new Squid developers to make steady progress.

And just to avoid misunderstand, this change request is _not_ limited to reporting. I suspect problematic `/etc/host` should be treated as a fatal configuration problem. Admins can use hosts_file to supply a working alternative if they cannot fix the original.